core/block_validator.go

```diff
@@ -99,6 +112,30 @@ func (v *BlockValidator) ValidateState(block *types.Block, statedb *state.StateD
 	if root := statedb.IntermediateRoot(v.config.IsEIP158(header.Number())); header.Root() != root {
 		return fmt.Errorf("invalid merkle root (remote: %x local: %x)", header.Root(), root)
 	}
+	// Collect ETXs emitted from each successful transaction
+	var emittedEtxs types.Transactions
+	for _, receipt := range receipts {
+		if receipt.Status == types.ReceiptStatusSuccessful {
+			for _, etx := range receipt.Etxs {
+				emittedEtxs = append(emittedEtxs, etx)
+			}
+		}
+	}
+	// Confirm the ETXs emitted by the transactions in this block exactly match the
+	// ETXs given in the block body
+	if etxHash := types.DeriveSha(emittedEtxs, trie.NewStackTrie(nil)); etxHash != header.EtxHash() {
+		return fmt.Errorf("invalid etx hash (remote: %x local: %x)", header.EtxHash(), etxHash)
+	}
+	// Collect the ETX rollup with emitted ETXs since the last coincident block,
+	// excluding this block.
+	etxRollup, err := v.hc.CollectEtxRollup(block)
```

Comment:
Its a tradeoff between storage access time & compute time. Each of the elements accessed in a rollup (headers & block bodies) are cached, so the access time is very quick this way (in steady state). The recursion essentially just parses the tree of objects to collect into the rollup.

Re-caching the entire rollup set at each block would have the following tradeoffs compared to this approach:
Pros:
* no recursion -- marginally easier to understand
Cons:
* significant data duplication (90% of each entry will be duplicate data of its parent entry)
* potentially slower data access times, since these new data objects will not have been cached before (requires a new db write of a very large object) and cannot live in cache long at all (because they are huge), so future accesses are likely to require a db read (again of a very large object)

In comparison, this approach has high probability of cache hits, and the occasional cache miss will be for a much smaller data object.

Reply:
ok. fine as is.